File: 9.cpp


Cpp code:
# include<iostream>
# include<string.h>
# include<stdlib.h>
using namespace std;
class node
{
public: char data;
int r;
node *left,*right;
node()
{
left=right=NULL;
}
node(int x)
{
data=x;
left=right=NULL;
}
};
class rstack
{
char reg[10][4];
int top;
public:
rstack()
{
top=-1;
}
void init(int k,char *name);
void swap()
{
char name[10];
strcpy(name,reg[top]);
strcpy(reg[top],reg[top-1]);
strcpy(reg[top-1],name);
}
char *pop()
{
return(reg[top--]);
}
int topv()
{
return top;
}
void push(char *tr)
{
strcpy(reg[++top],tr);
}
char *topd()
{
return (reg[top]);
}
};
void rstack::init(int k,char *name)
{ char temp[3];
for(int i=k-1;i>=0;i--)
{
strcpy(reg[++top],name);
//itoa(i,temp,10);
snprintf(temp, 10, "%d", i);
strcat(reg[top],temp);
}
}
class tree
{
node *root;
rstack m,tstack;
int rg;
public:
tree(){root=NULL;}
node *create1();
void disp(char c);
void preorder(node *t);
void create() {root=create1();}
void preorder() {preorder(root);}
void label(node * t);
void label(){label(root);}
void gencode(node *t);
void gencode()
{
cout<<"enter regs\n";
cin>>rg;
m.init(rg,"R");
tstack.init(6,"T");
gencode(root);
}
};
void tree::disp(char c)
{
switch(c)
{
case '+' : cout<<"ADD ";
break;
case '*' :cout<<"MUL ";
break;
case '-':cout<<"SUB ";
break;
case '/':cout<<"DIV ";
break;
}
}
void tree::gencode(node *t)
{ char *name,*name1;
/*case 0*/
if(t->r==1&&t->left==NULL&&t->right==NULL)
{
cout<<"MOV "<<t->data<<","<<m.topd()<<"\n";;
}
/*case 1*/
else
if(t->left!=NULL&&t->right->r==0)
{
gencode(t->left);
disp(t->data);
cout<<t->right->data<<" "<<m.topd()<<"\n";
}
/*case 2*/
else
if((t->left->r<t->right->r)&&(t->left->r<rg))
{
// cout<<"case 2\n";
m.swap();
gencode(t->right);
name=m.pop();
// cout<<"popped reg "<<name<<"\n"<<"top= "<<m.topv()<<"\n";
gencode(t->left);
disp(t->data);
cout<<name<<" " <<","<<m.topd()<<"\n";
m.push(name);
m.swap();
}
/*case 3*/
else if((t->right->r<=t->left->r)&& (t->right->r<rg))
{
// cout<<"case 3\n";
gencode(t->left);
name=m.pop();
// cout<<"popped reg "<<name<<"\n"<<"top= "<<m.topv();
gencode(t->right);
disp(t->data);
cout<<m.topd()<<","<<name<<"\n";
m.push(name);
}
else
{
gencode(t->right);
name=tstack.pop();
cout<<"MOV "<<m.topd()<<","<<name<<"\n";
gencode(t->left);
disp(t->data);
cout<<name<<","<<m.topd()<<"\n";
tstack.push(name);
}
}
void tree::label(node *t)
{
if(t->left->left==NULL&&t->left->right==NULL)
{
t->left->r=1;
}
if(t->left->left!=NULL||t->left->right!=NULL)
label(t->left);
if(t->right->left==NULL&&t->right->right==NULL)
{
t->right->r=0;
}
if(t->right->left!=NULL||t->right->right!=NULL)
label(t->right);
if(t->left!=NULL&&t->right!=NULL)
{
if(t->left->r>t->right->r)
t->r=t->left->r;
else if(t->right->r>t->left->r)
t->r=t->right->r;
else
t->r=t->left->r+1;
}
}
node *tree::create1()
{
node *p;
char x;
cout<<"Enter node data\n";
cin>>x;
if(x=='Z')
return NULL;
else
{
p=new node(x);
cout<<"Enter left child of "<<x<<endl;
p->left=create1();
cout<<"Enter right child of "<<x<<endl;
p->right=create1();
}
return p;
}
void tree::preorder(node *t)
{
if(t)
{
cout<<t->data<<"\t"<<"(label="<<t->r<<")\t";
preorder(t->left);
preorder(t->right);
}
}
int main()
{
tree t;
int choice,n;
char ch;
do
{
cout<<"******MENU******\n";
cout<<"case 1:create tree\n";
cout<<"case 2:preorder\n";
cout<<"case 3:label\n";
cout<<"case 4:gencode\n";
cout<<"Enter your choice\n";
cin>>choice;
switch(choice)
{
case 1: t.create();
break;
case 2:t.preorder();
break;
case 3:t.label();
break;
case 4:t.gencode();
break;
default:exit(0);
}
cout<<"\nDo you want to continue\n";
cin>>ch;
}while(ch=='y');
return 0;
}

File: 9.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.r = 0
        self.left = None
        self.right = None

class RStack:
    def __init__(self):
        self.reg = [""] * 10
        self.top = -1
    
    def init(self, k, name):
        for i in range(k-1, -1, -1):
            self.top += 1
            self.reg[self.top] = name + str(i)

    def swap(self):
        name = self.reg[self.top]
        self.reg[self.top] = self.reg[self.top-1]
        self.reg[self.top-1] = name

    def pop(self):
        self.top -= 1
        return self.reg[self.top+1]

    def topv(self):
        return self.top

    def push(self, tr):
        self.top += 1
        self.reg[self.top] = tr

    def topd(self):
        return self.reg[self.top]

class Tree:
    def __init__(self):
        x = input("Enter root node: ")
        self.root = self.create(x)
        self.m = RStack()
        self.tstack = RStack()
        self.rg = 0

        
    def create(self, x):
        if x == '':
            return None
        else:
            p = Node(x)
            l = input(f"Enter left child of {x} : ")
            p.left = self.create(l)
            r = input(f"Enter right child of {x} :")
            p.right = self.create(r)
        return p

    def disp(self, c):
        if c == '+':
            print("ADD", end=" ")
        elif c == '*':
            print("MUL", end=" ")
        elif c == '-':
            print("SUB", end=" ")
        elif c == '/':
            print("DIV", end=" ")

    def gencode(self, t):
        if t.r == 1 and t.left is None and t.right is None:
            print("MOV", t.data, ",", self.m.topd())
        elif t.left is not None and t.right.r == 0:
            self.gencode(t.left)
            self.disp(t.data)
            print(t.right.data, self.m.topd())
        elif t.left.r < t.right.r and t.left.r < self.rg:
            self.m.swap()
            self.gencode(t.right)
            name = self.m.pop()
            self.gencode(t.left)
            self.disp(t.data)
            print(name, ",", self.m.topd())
            self.m.push(name)
            self.m.swap()
        elif t.right.r <= t.left.r and t.right.r < self.rg:
            self.gencode(t.left)
            name = self.m.pop()
            self.gencode(t.right)
            self.disp(t.data)
            print(self.m.topd(), ",", name)
            self.m.push(name)
        else:
            self.gencode(t.right)
            name = self.tstack.pop()
            print("MOV", self.m.topd(), ",", name)
            self.gencode(t.left)
            self.disp(t.data)
            print(name, ",", self.m.topd())
            self.tstack.push(name)

    def label(self, t):
        if t.left.left is None and t.left.right is None:
            t.left.r = 1
        if t.left.left is not None or t.left.right is not None:
            self.label(t.left)
        if t.right.left is None and t.right.right is None:
            t.right.r = 0
        if t.right.left is not None or t.right.right is not None:
            self.label(t.right)
        if t.left is not None and t.right is not None:
            if t.left.r > t.right.r:
                t.r = t.left.r
            elif t.right.r > t.left.r:
                t.r = t.right.r
            else:
                t.r = t.left.r + 1

    def preorder(self, t):
        if t:
            print(t.data, "\t(label=", t.r, ")\t")
            self.preorder(t.left)
            self.preorder(t.right)

t = Tree()
print("\nTree: ")
t.preorder(t.root)
print("\nLabeled tree: ")
t.label(t.root)
t.preorder(t.root)
t.rg = int(input("\nEnter number of registers: "))
t.m.init(t.rg, "R")
t.tstack.init(6, "T")
print("\nGenerated code: ")
t.gencode(t.root)
